File: src/main/java/bbarena/model/event/game/ScatterBallEvent.java
package bbarena.model.event.game;

import bbarena.model.Arena;
import bbarena.model.Coordinate;
import bbarena.model.Direction;
import bbarena.model.pitch.Ball;
import bbarena.model.pitch.BallMove;
import bbarena.model.pitch.SquareDestination;
import bbarena.model.util.Concat;
import bbarena.model.util.Pair;

public class ScatterBallEvent extends BallEvent {

	private static final long serialVersionUID = -6262949311066708660L;

	private Direction _direction = null;
	private Integer _distance = 1;
	private Coordinate _origin;
	private SquareDestination _destination;
	private BallMove.BallMoveType _type = BallMove.BallMoveType.SCATTER;

	public ScatterBallEvent(int ballId) {
		super(ballId);
	}

	public ScatterBallEvent(int ballId, int direction) {
		this(ballId, direction, 1);
	}
	public ScatterBallEvent(int ballId, int direction, int distance) {
		super(ballId);
		_direction = Direction.getDirection(direction);
		_distance = distance;
	}

	public BallMove.BallMoveType getType() {
		return _type;
	}

	public void setType(BallMove.BallMoveType type) {
		_type = type;
	}

	public void setBall(Ball ball) {
        ball = ball;
    }

	public void setDirection(Direction direction) {
		_direction = direction;
	}

	public Integer getDistance() {
		return _distance;
	}

	public void setDistance(Integer distance) {
		_distance = distance;
	}

	public Coordinate getOrigin() {
		return _origin;
	}

	public SquareDestination getDestination() {
		return _destination;
	}

	@Override
	protected void doEvent(Arena arena) {
        Ball ball = getBall(arena);

        _origin = ball.getSquare().getCoords();
        _destination = arena.getPitch().getDestination(_origin, _direction, _distance);
        arena.getPitch().moveBall(ball, _direction, _destination.getEffectiveDistance(), _type);
    }

	@Override
	protected void undoEvent(Arena arena) {
        Ball ball = getBall(arena);
        arena.getPitch().moveBall(ball, _direction.inverse(), _destination.getEffectiveDistance(), _type);
    }

	@Override
	public String getString() {
		return Concat.buildLog(getClass(),
				new Pair("ballId", getBallId()),
				new Pair("direction", _direction),
				new Pair("distance", _distance));
	}

	public Direction getDirection() {
		return _direction;
	}

}

File: src/main/kotlin/bbarena/model/event/game/ScatterBallEvent.kt
package bbarena.model.event.game

import bbarena.model.Arena
import bbarena.model.Coordinate
import bbarena.model.Direction
import bbarena.model.pitch.Ball
import bbarena.model.pitch.BallMove
import bbarena.model.pitch.SquareDestination
import bbarena.model.util.Concat

open class ScatterBallEvent : BallEvent {

    var direction: Direction? = null
    var distance: Int? = 1
    var origin: Coordinate? = null
        private set
    var destination: SquareDestination? = null
        private set
    var type: BallMove.BallMoveType = BallMove.BallMoveType.SCATTER

    constructor(ballId: Int) : super(ballId) {}
    @JvmOverloads
    constructor(ballId: Int, direction: Int, distance: Int = 1) : super(ballId) {
        this.direction = Direction.getDirection(direction)
        this.distance = distance
    }

    fun setBall(ball: Ball) {
        var ball = ball
        ball = ball
    }

    override fun doEvent(arena: Arena) {
        val ball = getBall(arena)

        origin = ball.square.coords
        destination = arena.pitch.getDestination(origin, direction, distance!!)
        arena.pitch.moveBall(ball, direction, destination!!.effectiveDistance!!, type)
    }

    override fun undoEvent(arena: Arena) {
        val ball = getBall(arena)
        arena.pitch.moveBall(ball, direction!!.inverse(), destination!!.effectiveDistance!!, type)
    }

    override fun getString(): String {
        return Concat.buildLog(javaClass,
                Pair("ballId", ballId),
                Pair("direction", direction),
                Pair("distance", distance))
    }

    companion object {

        private val serialVersionUID = -6262949311066708660L
    }

}
